README.md
# Arduino Si5351 Library tuned for size and click noise free #

This library is tuned for size on the Arduino platform, it will control CLK0, CLK1 and CLK2 outputs for the Si5351A (the version with just 3 clocks out, but you will not be able to use the three at once).

## Inspiration ##

This work is based on the previous work of these great people:

* [Etherkit/NT7S:](https://github.com/etherkit/Si5351Arduino) The mainstream full featured lib, with big code as well (based on Linux kernel code)
* [QRP Labs demo code from Hans Summers:](http://qrp-labs.com/synth/si5351ademo.html) The smallest and simple ones on the net
* [DK7IH demo code:](https://radiotransmitter.wordpress.com/category/si5351a/) The first clickless noise code on the wild
* [Jerry Gaffke](https://github.com/afarhan/ubitx.git) integer routines for the Raduino and ubitx

## Set your Goal and make an estrategy ##

There is a few routines in the Internet to manage the Si5351 chip, all of them has a few distinct feature set because they use different strategies (different goals) that make them unique in some way.

My goal is this:

* Keep it as small as possible (Smallest firmware footprint)
* Less phase and click noise possible (Playing with every trick possible)

The main purpose is to be used in Radio receiver projects, so this two mentioned goals are the golden rule.

Let's list some of goals achievements and bonuses:

**Small firmware footprint:**

A basic sketch to set just only one clock out to a given frequency with a change in power and correcting the XTAL ppm error is only ~3.3 kBytes of firmware (~10% of an Arduino Uno)

The same settings with the [Si5351Arduino library (Etherkit)](https://github.com/etherkit/Si5351Arduino) will give you a bigger firmware space of ~10 kBytes or 31% of an Arduino Uno.

[Jerry Gaffke](https://github.com/afarhan/ubitx.git) embedded routines in the ubitx transceiver has the smallest footprint in the Arduino platform I have seen, but has worst phase noise and smallest frequency range.

**Phase noise to the minimum:**

We use every trick on the datasheet, OEM Application Notes or the Internet to minimize phase noise. (Even a few ones learned on the process)

For example the [Etherkit](https://github.com/etherkit/Si5351Arduino) library and [Jerry Gaffke](https://github.com/afarhan/ubitx.git) embedded routines uses some but not all the tricks to minimize phase noise (Etherkit one gives control over all features, Jerry Gaffke has small footprint and in the process he sacrifice phase noise and frequency range)

**Click noise free:**

If you play from the book (Datasheet and Application Notes) you will have a "click-like" noise burst every time you change the output frequency.

That's not a problem if you intend to use only fixed frequencies at the output, but if you plan to sweep or use it on any application that moves the frequency that click-like noise will haunt you. (like in a radio receiver or transceiver)

I have learned a few tricks from many sources in the Internet and after some local testing I have came across a scheme that make this lib virtually click-noise-less; see the "Click noise free" section below for details.

**Fast frequency changes:**

This was a side effect of the last trick to minimize the click noise, see the "Click noise free" section below for details.

Summary: other routines write all registers for every frequency change, I write half of them most of the time, speeding up the process.

**Two of three**

Yes, there is no such thing as free lunch, to get all the above features and the ones that follow I have to make some sacrifices, in this case spare one of the outputs. See "Two of three" section below.

## Features ##

This are so far the implemented features (Any particular wish? use the Issues tab for that):

* Custom XTAL passing on init (Default is 27.000 MHz (See _Si.init()_ )
* You can pass a correction to the xtal while running (See _Si.correction()_ )
* You have a fast way to power off all outputs of the Chip at once. (See _Si.off()_ )
* You can enable/disable any output at any time (See _Si.enable(clk) and Si.disable(clk)_ )
* By default all outputs are off after the Si.init() procedure. You has to enable them by hand.
* You can only have 2 of the 3 outputs running at any moment (See "Two of three" section below)
* Power control on each output independently (See _Si.setPower(clk, level)_ on the lib header)
* Initial power defaults to the lowest level (2mA) for all outputs.
* You don't need to include and configure the Wire (I2C) library, this lib do that for you already.
* Frequency limits are not hard coded on the lib, so you can stress your hardware to it's particular limit (_You can move usually from ~3kHz to ~225 MHz, far away from the 8kHz to 160 MHz limits from the datasheet_)
* You has a way to verify the status of a particular clock (_Enabled/Disabled by the Si.clkOn[clk] var_)
* From v0.5 and beyond we saved more than 1 kbyte of your precious firmware space due to the use of all integer math now (Worst induced error is below +/- 1 Hz)
* Overclock, yes, you can move the limits upward up to ~250MHz (see the "OVERCLOCK" section below)
* Improved the click noise algorithm to get even more click noise reduction (see Click noise free section below)
* Fast frequency changes as part of the improved click noise algorithm (see Click noise free section below)

## How to use the lib ##

Get the lib by cloning this git repository or get it by clicking the green "Download button" on the page.

Move it or extract it on your library directory

Include the lib in your code:


```
(... your code here ...)

// now include the library
#include "si5351mcu.h"

// lib instantiation as "Si"
Si5351mcu Si;

(... more of your code here ...)

```

Follow this sequence on you setup() procedure:

* Initialize the library with the default or optional Xtal Clock.
* Apply correction factor (if needed)
* Set some frequencies to the desired outputs.
* Enable the desired outputs

Here you have an example code ("Si" is the lib instance):

```
setup() {
    (... your code here ...)

    //////////////////////////////////
    //        Si5351 functions       /
    //////////////////////////////////

    // Init the library, in this case with the default 27.000 Mhz Xtal
    Si.init();

    // commented Init procedure for a not default 25.000 MHz xtal
    //Si.init(25000000L);

    // Optional, apply a pre-calculated correction factor
    Si.correction(-150);    // Xtal is low by 150 Hz

    // Enable the desired outputs with some frequencies
    Si.setFreq(0, 25000000);       // CLK0 output 25.000 MHz
    Si.setFreq(1, 145000000);      // CLK1 output 145.000 MHz

    // enable the outputs
    Si.enable(0);
    Si.enable(1);

    (... more of your code here ...)
}

```

If you need to apply/vary the correction factor **after** the setup process you will get a click noise on the next setFreq() to apply the changes.

Use it, you can enable, disable, change the power or frequency, see this code fragment with some examples:

```
loop() {
    (... your code here ...)

    // disable clk1
    Si.disable(1);

    // change the power of clk0 to 4mA
    Si.setPower(0, SIOUT_4mA);

    // apply a correction factor of 300 Hz (correction will be applied on the next Si.setFreq() call)
    Si.correction(300);

    // change the clk0 output frequency
    Si.setFreq(0, 7110000);

    // power of all outputs
    Si.off();

    (... more of your code here ...)
}
```


## OVERCLOCK ##

Yes, you can overclock the Si5351, the datasheet states that the VCO moves from 600 to 900 MHz and that gives us a usable range from ~3 kHz to 225 MHz.

But what if we can move the VCO frequency to a higher values?

The overclock feature does just that, use a higher top limit for the VCO on the calculations. In my test with two batch of the Si5351A I can get safely up to 1.000 GHz without trouble; in one batch the PLL unlocks around 1.1 GHz and in the other about 1.050 GHz; so I recommend not going beyond 1.000 GHz.

With a maximum VCO of 1.000 GHz and a lower division factor of 4 we have jumped from a 225 MHz to 250 MHz top frequency that can be generated with our cheap chip.

**Some "must include" WARNINGS:**

* The chip was not intended to go that high, so, use it with caution and test it on your hardware moving the overclock limit in steps of 10 MHz starting with 900 MHz and testing with every change until it does not work; then go down by 10 MHz to be in a safe zone.
* Moving the upper limit has its penalty on the lower level, your bottom frequency will move from the ~3 kHz to ~10 kHz range.
* The phase noise of the output if worst as you use a higher frequency, at a _**fixed**_ 250 MHz it's noticeable but no so bad for a TTL or CMOS clock application.
* The phase noise is specially bad if you make a sweep or move action beyond 180 MHz; the phase noise from the unlock state to next lock of the PLL is very noticeable in a spectrum analyzer, even on a cheap RTL-SDR one.
* I recommend to only use the outputs beyond 150 MHz as fixed value and don't move them if you cares about phase noise.

**How to do it?**

You need to declare a macro with the overclock value **BEFORE** the library include, just like this:

```
(... your code here ...)

// Using the overclock feature for the Si5351mcu library
#define SI_OVERCLOCK 1000000000L      // 1.0 GHz in Hz

// now include the library
#include "si5351mcu.h"

// lib instantiation as "Si"
Si5351mcu Si;

// now you can generate frequencies from ~10 kHz up to 250 MHz.

(... more of your code here ...)

```

## Click noise free ##

Click-like noise came from a few sources as per my testing:

* Turn off then on the CLKx outputs (Register 3. Output Enable Control)
* Power down then on the Msynths (CLKx_PDN bits for every Msynth)
* Reset the PLL (Register 177: PLL Reset)

We are concerned about click noise only when changing from one frequency to the other, so if we don't touch the output enable control or the power down msynth registers once activated; then we are set to avoid click from this two first sources.

The last one is tricky, in theory a PLL does not need to be reseted every time you change it's output frequency as it's a **P**hase **L**ocked **L**oop and it's a self correcting algorithm/hardware.

That last idea was put to test on a simple scheme: what if I set a fixed output divider Msynth and move the VCO for it's entire range without resetting it on any point?

If the "PLL reset" is in deed needed I will have some strange behavior at some point, right?

But practice confirmed my idea, I can set a output Msynth of 6 and move the VCO (PLL) for the entire range (600 to 900 MHz) and get a continuous and stable output from 100 to 150 MHz.

Then what is for the "PLL reset" (Register 177) in practice?

Some further test showed that this "reset" function is applied no to the PLL Msynth, but to the output Msynth and not in every case, yes, it has a bad name or a bad explained name.

After some test I find that you need the "PLL reset" (Register 177) trick only on some cases when you change the value of the output divider Msynth.

Implementing that in code was easy, an array to keep track of the actual output divider Msynth and only write it to the chip and reset "the PLL" when it's needed.

Hey! that leads to a I2C time reduction by half (most of the time) as a side effect!

Most of the time when you are making a sweep the output divider Msynth has a constant value and you only moves the VCO (PLL) Then I wrote just 8 bytes to the I2C bus (to control the VCO/PLL) instead of 16 (8 for the VCO/PLL & 8 more for the output divider Msynth) or 17 (16 + reset byte) most of the time, cutting time between writes to half making frequency changes 2x fast as before.

## Two of three ##

Yes, there is a tittle catch here with CLK1 and CLK2: both share PLL_B and as we use our own algorithm to calculate the frequencies and minimize phase noise you can only use one of them at a time.

Note: _In practice you can, but the other will move from the frequency you set, which is an unexpected behavior, so I made them mutually exclusive (CLK1 and CLK2)._

This are the valid combinations for independent clocks output.

* CLK0 and CLK1
* CLK0 and CLK2

Again: You can't use CLK1 and CLK2 at the same time, as soon as you set one of them the other will shut off. That's why you get two of three and one of them must be always CLK0.

## Author & contributors ##

The only author is Pavel Milanes, CO7WT, a cuban amateur radio operator; reachable at pavelmc@gmail.com, Until now I have no contributors or sponsors.

## Where to download the latest version? ##

Always download the latest version from the [github repository](https://github.com/pavelmc/Si5351mcu/)

See ChangeLog.md and Version files on this repository to know what are the latest changes and versions.

## If you like to give thanks... ##

No payment of whatsoever is required to use this code: this is [Free/Libre Software](https://en.wikipedia.org/wiki/Software_Libre), nevertheless donations are very welcomed.

I live in Cuba island and the Internet is very expensive here (USD $10.00 per 1GB of data), you can donate anonymously internet time or cell mobile data to me via [Ding Topups](https://www.ding.com/) to keep me connected and developing for the homebrew/DIY community.

If you like to do so, please go to Ding, select Cuba, select Cubacel and use my number `(+53) 538-478-19` (feel free to call me if you like)

Thanks!


ChangeLog.md
# Si5351mcu Changelog File #

## v0.6.2 (January 21, 2020) ##

* Fix a type mismatch that causes an error when you compile it under ESP8266/ESP32

## v0.6.1 (Augus 1, 2018) ##

* Documentation update and completion to match code features and tricks

## v0.6 (February 10, 2018) ##

* Feature: you can now OVERCLOCK the Si5351 to get up to 250 MHz from it, see the overclock subject on the README.md
* Documentation improvements and re-arrangement.

## v0.5 (February 7, 2018) ##

* Feature: All integer math now, induced error must be at worst +/- 2 Hz.
* Feature: Clock status via clkOn[clk] public var.
* Bug Fix: Output divider low limit safe guard in place (it make some trouble under some circumstances)
* New super simple example.

## v0.4 (August 2, 2017) ##

* Bug Fix: Triaged a strange level problem with freqs above ~112 Mhz, the signal level start dropping around 112MHz and ~150 MHz suddenly go beyond limits (high) to slowly drop again up to the end. Fact: the lib needs a reset() on EVERY frequency change above VCO/8 (~112MHz). Remember that datasheet specs are 8KHz to 160MHz, but we are pushing it up to ~225 Mhz (max_vco/4)
* Code refractory on some points to match correct behavior, as previous code has little bugs introduced by bad documentation from Silicon Labs (+1 for the chip; -5 for the docs errors). See Bitx20 mail-lits archives for June-August 2017 for more info and the debate.

## v0.3 (June 14, 2017) ##

* Feature: the lib now handle the include and start of the I2C (Wire) library internally via the init procedures
* Added a new generic init() procedure to handle the default parameters
* The init() function is required from now on (MANDATORY)
* Fixed the way we handled the base xtal and the correction factor

## v0.2rc (April 23, 2017) ##

* Added power level support for each output independently, watch out!: setting the power level will enable the output.
* Set default power to the lowest one (2mA) from the maximun possible (8mA).
* Fixed the need for a reset after each correction, it does it now automatically
* Added a init function to pass a custom xtal
* Modified the example to show no need for a init unless you use a different xtal
* Improved the keywords.txt file to improve edition on the Arduino IDE
* Included a "features" section on the README.md

## v0.1rc (April 20, 2017) ##

* Added enable(), disable() and off() functions.
* Added support for handling all the three outputs of the Si5351A, (CLK1 & CLK2 are mutually-exclusive)
* Updated the example with the new functions.
* Improved library logic by reusing and optimizing functions.
* Improved the documentation and comments (lib code, README and example)
* The compiled code is slightly smaller now (~1% on an ATMega328p)
* Added Changelog and version files.
* Extensive tests made to validate every function.

## Initial Release, v0.0 (April 9, 2017) ##

* Basic functionality.


src/si5351mcu.h
/*
 * si5351mcu - Si5351 library for Arduino MCU tuned for size and click-less
 *
 * Copyright (C) 2017 Pavel Milanes <pavelmc@gmail.com>
 *
 * Many chunk of codes are derived-from/copied from other libs
 * all GNU GPL licenced:
 *  - Linux Kernel (www.kernel.org)
 *  - Hans Summers libs and demo code (qrp-labs.com)
 *  - Etherkit (NT7S) Si5351 libs on github
 *  - DK7IH example.
 *  - Jerry Gaffke integer routines for the bitx20 group
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


/****************************************************************************
 *  This lib tricks:
 *
 * CLK0 will use PLLA
 * CLK1 will use PLLB
 * CLK2 will use PLLB
 *
 * Lib defaults
 * - XTAL is 27 Mhz.
 * - Always put the internal 8pF across the xtal legs to GND
 * - lowest power output (2mA)
 * - After the init all outputs are off, you need to enable them in your code.
 *
 * The correction procedure is not for the PPM as other libs, this
 * is just +/- Hz to the XTAL freq, you may get a click noise after
 * applying a correction
 *
 * The init procedure is mandatory as it set the Xtal (the default or a custom
 * one) and prepare the Wire (I2C) library for operation.
 ****************************************************************************/

#ifndef SI5351MCU_H
#define SI5351MCU_H

// rigor includes
#include "Arduino.h"
#include "Wire.h"

// default I2C address of the Si5351
#define SIADDR 0x60

// register's power modifiers
#define SIOUT_2mA 0
#define SIOUT_4mA 1
#define SIOUT_6mA 2
#define SIOUT_8mA 3

// registers base (2mA by default)
#define SICLK0_R   76       // 0b1001100
#define SICLK12_R 108       // 0b1101100


class Si5351mcu {
    public:
        // default init procedure
        void init(void);

        // custom init procedure (XTAL in Hz);
        void init(uint32_t);

        // reset all PLLs
        void reset(void);

        // set CLKx(0..2) to freq (Hz)
        void setFreq(uint8_t, uint32_t);

        // pass a correction factor
        void correction(int32_t);

        // enable some CLKx output
        void enable(uint8_t);

        // disable some CLKx output
        void disable(uint8_t);

        // disable all outputs
        void off(void);

        // set power output to a specific clk
        void setPower(uint8_t, uint8_t);

        // var to check the clock state
        bool clkOn[3] = {0, 0, 0};


    private:
        // used to talk with the chip, via Arduino Wire lib
        void i2cWrite(uint8_t, uint8_t);

        // base xtal freq, over this we apply the correction factor
        // by default 27 MHz
        uint32_t base_xtal = 27000000L;

        // this is the work value, with the correction applied
        // via the correction() procedure
        uint32_t int_xtal = base_xtal;

        // clk# power holders (2ma by default)
        uint8_t clkpower[3] = {0, 0, 0};

        // local var to keep track of when to reset the "pll"
        /*******************************************************
         * BAD CONCEPT on the datasheet and AN:
         *
         * The chip has a soft-reset for PLL A & B but in
         * practice the PLL does not need to be reseted.
         *
         * Test shows that if you fix the Msynth output
         * dividers and move any of the VCO from bottom to top
         * the frequency moves smooth and clean, no reset needed
         *
         * The reset is needed when you changes the value of the
         * Msynth output divider, even so it's not always needed
         * so we use this var to keep track of all three and only
         * reset the "PLL" when this value changes to be sure
         *
         * Its a word (16 bit) because the final max value is 900
         *******************************************************/
        uint16_t omsynth[3] = {0, 0, 0};
};


#endif //SI5351MCU_H


src/si5351mcu.cpp
/*
 * si5351mcu - Si5351 library for Arduino MCU tuned for size and click-less
 *
 * Copyright (C) 2017 Pavel Milanes <pavelmc@gmail.com>
 *
 * Many chunk of codes are derived-from/copied from other libs
 * all GNU GPL licenced:
 *  - Linux Kernel (www.kernel.org)
 *  - Hans Summers libs and demo code (qrp-labs.com)
 *  - Etherkit (NT7S) Si5351 libs on github
 *  - DK7IH example.
 *  - Jerry Gaffke integer routines for the bitx20 group
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "Arduino.h"
#include "si5351mcu.h"

// wire library loading, if not defined
#ifndef WIRE_H
    #include "Wire.h"
#endif

/*****************************************************************************
 * This is the default init procedure, it set the Si5351 with this params:
 * XTAL 27.000 Mhz
 *****************************************************************************/
 void Si5351mcu::init(void) {
    // init with the default freq
    init(int_xtal);
}


/*****************************************************************************
 * This is the custom init procedure, it's used to pass a custom xtal
 * and has the duty of init the I2C protocol handshake
 *****************************************************************************/
 void Si5351mcu::init(uint32_t nxtal) {
    // set the new base xtal freq
    base_xtal = int_xtal = nxtal;

    // start I2C (wire) procedures
    Wire.begin();

    // power off all the outputs
    off();
}


/*****************************************************************************
 * This function set the freq of the corresponding clock.
 *
 * In my tests my Si5351 can work between 7,8 Khz and ~225 Mhz [~250 MHz with
 * overclocking] as usual YMMV
 *
 * Click noise:
 * - The lib has a reset programmed [aka: click noise] every time it needs to
 *   change the output divider of a particular MSynth, if you move in big steps
 *   this can lead to an increased rate of click noise per tunning step.
 * - If you move at a pace of a few Hz each time the output divider will
 *   change at a low rate, hence less click noise per tunning step.
 * - The output divider moves [change] faster at high frequencies, so at HF the
 *   clikc noise is at the real minimum possible.
 *
 * [See the README.md file for other details]
 ****************************************************************************/
void Si5351mcu::setFreq(uint8_t clk, uint32_t freq) {
    uint8_t a, R = 1, shifts = 0;
    uint32_t b, c, f, fvco, outdivider;
    uint32_t MSx_P1, MSNx_P1, MSNx_P2, MSNx_P3;

    // Overclock option
    #ifdef SI_OVERCLOCK
        // user a overclock setting for the VCO, max value in my hardware
        // was 1.05 to 1.1 GHz, as usual YMMV [See README.md for details]
        outdivider = SI_OVERCLOCK / freq;
    #else
        // normal VCO from the datasheet and AN
        // With 900 MHz beeing the maximum internal PLL-Frequency
        outdivider = 900000000 / freq;
    #endif

    // use additional Output divider ("R")
    while (outdivider > 900) {
        R = R * 2;
        outdivider = outdivider / 2;
    }

    // finds the even divider which delivers the intended Frequency
    if (outdivider % 2) outdivider--;

    // Calculate the PLL-Frequency (given the even divider)
    fvco = outdivider * R * freq;

    // Convert the Output Divider to the bit-setting required in register 44
    switch (R) {
        case 1:   R = 0; break;
        case 2:   R = 16; break;
        case 4:   R = 32; break;
        case 8:   R = 48; break;
        case 16:  R = 64; break;
        case 32:  R = 80; break;
        case 64:  R = 96; break;
        case 128: R = 112; break;
    }

    // we have now the integer part of the output msynth
    // the b & c is fixed below
    MSx_P1 = 128 * outdivider - 512;

    // calc the a/b/c for the PLL Msynth
    /***************************************************************************
    * We will use integer only on the b/c relation, and will >> 5 (/32) both
    * to fit it on the 1048 k limit of C and keep the relation
    * the most accurate possible, this works fine with xtals from
    * 24 to 28 Mhz.
    *
    * This will give errors of about +/- 2 Hz maximum
    * as per my test and simulations in the worst case, well below the
    * XTAl ppm error...
    *
    * This will free more than 1K of the final eeprom
    *
    ****************************************************************************/
    a = fvco / int_xtal;
    b = (fvco % int_xtal) >> 5;     // Integer par of the fraction
                                    // scaled to match "c" limits
    c = int_xtal >> 5;              // "c" scaled to match it's limits
                                    // in the register

    // f is (128*b)/c to mimic the Floor(128*(b/c)) from the datasheet
    f = (128 * b) / c;

    // build the registers to write
    MSNx_P1 = 128 * a + f - 512;
    MSNx_P2 = 128 * b - f * c;
    MSNx_P3 = c;

    // PLLs and CLK# registers are allocated with a shift, we handle that with
    // the shifts var to make code smaller
    if (clk > 0 ) shifts = 8;

    // plls, A & B registers separated by 8 bytes
    i2cWrite(26 + shifts, (MSNx_P3 & 65280) >> 8);   // Bits [15:8] of MSNx_P3 in register 26
    i2cWrite(27 + shifts, MSNx_P3 & 255);
    i2cWrite(28 + shifts, (MSNx_P1 & 196608) >> 16);
    i2cWrite(29 + shifts, (MSNx_P1 & 65280) >> 8);   // Bits [15:8]  of MSNx_P1 in register 29
    i2cWrite(30 + shifts, MSNx_P1 & 255);            // Bits [7:0]  of MSNx_P1 in register 30
    i2cWrite(31 + shifts, ((MSNx_P3 & 983040) >> 12) | ((MSNx_P2 & 983040) >> 16)); // Parts of MSNx_P3 and MSNx_P1
    i2cWrite(32 + shifts, (MSNx_P2 & 65280) >> 8);   // Bits [15:8]  of MSNx_P2 in register 32
    i2cWrite(33 + shifts, MSNx_P2 & 255);            // Bits [7:0]  of MSNx_P2 in register 33

    // Write the output divider msynth only if we need to, in this way we can
    // speed up the frequency changes almost by half the time most of the time
    // and the main goal is to avoid the nasty click noise on freq change
    if (omsynth[clk] != outdivider) {
        // CLK# registers are exactly 8 * clk# bytes shifted from a base register.
        shifts = clk * 8;

        // multisynths
        i2cWrite(42 + shifts, 0);                        // Bits [15:8] of MS0_P3 (always 0) in register 42
        i2cWrite(43 + shifts, 1);                        // Bits [7:0]  of MS0_P3 (always 1) in register 43
        // See datasheet, special trick when R=4
        if (outdivider == 4) {
            i2cWrite(44 + shifts, 12 | R);
            i2cWrite(45 + shifts, 0);            // Bits [15:8] of MSx_P1 must be 0
            i2cWrite(46 + shifts, 0);            // Bits [7:0] of MSx_P1 must be 0
        } else {
            i2cWrite(44 + shifts, ((MSx_P1 & 196608) >> 16) | R);  // Bits [17:16] of MSx_P1 in bits [1:0] and R in [7:4]
            i2cWrite(45 + shifts, (MSx_P1 & 65280) >> 8);    // Bits [15:8]  of MSx_P1 in register 45
            i2cWrite(46 + shifts, MSx_P1 & 255);             // Bits [7:0]  of MSx_P1 in register 46
        }
        i2cWrite(47 + shifts, 0);                        // Bits [19:16] of MS0_P2 and MS0_P3 are always 0
        i2cWrite(48 + shifts, 0);                        // Bits [15:8]  of MS0_P2 are always 0
        i2cWrite(49 + shifts, 0);                        // Bits [7:0]   of MS0_P2 are always 0

        // must reset the so called "PLL", in fact the output msynth
        reset();

        // keep track of the change
        omsynth[clk] = (uint16_t)outdivider;
    }
}


/*****************************************************************************
 * Reset of the PLLs and multisynths output enable
 *
 * This must be called to soft reset the PLLs and cycle the output of the
 * multisynths: this is the "click" noise source in the RF spectrum.
 *
 * So it must be avoided at all costs, so this lib just call it at the
 * initialization of the PLLs and when a correction is applied
 *
 * If you are concerned with accuracy you can implement a reset every
 * other Mhz to be sure it get exactly on spot.
 ****************************************************************************/
void Si5351mcu::reset(void) {
    // This soft-resets PLL A & B (32 + 128) in just one step
    i2cWrite(177, 0xA0);
}


/*****************************************************************************
 * Function to disable all outputs
 *
 * The PLL are kept running, just the m-synths are powered off.
 *
 * This allows to keep the chip warm and exactly on freq the next time you
 * enable an output.
 ****************************************************************************/
void Si5351mcu::off(void) {
    // This disable all the CLK outputs
    for (byte i=0; i<3; i++) disable(i);
}


/*****************************************************************************
 * Function to set the correction in Hz over the Si5351 XTAL.
 *
 * This will call a reset of the PLLs and multi-synths so it will produce a
 * click every time it's called
 ****************************************************************************/
void Si5351mcu::correction(int32_t diff) {
    // apply some corrections to the xtal
    int_xtal = base_xtal + diff;

    // reset the PLLs to apply the correction
    reset();
}


/*****************************************************************************
 * This function enables the selected output
 *
 * Beware: ZERO is clock output enabled
 *****************************************************************************/
void Si5351mcu::enable(uint8_t clk) {
    // var to handle the mask of the registers value
    uint8_t m = SICLK0_R;
    if (clk > 0) m = SICLK12_R;

    // write the register value
    i2cWrite(16 + clk, m + clkpower[clk]);

    // 1 & 2 are mutually exclusive
    if (clk == 1) disable(2);
    if (clk == 2) disable(1);

    // update the status of the clk
    clkOn[clk] = 1;
}


/*****************************************************************************
 * This function disables the selected output
 *
 * Beware: ONE is clock output disabled
 * *****************************************************************************/
void Si5351mcu::disable(uint8_t clk) {
    // send
    i2cWrite(16 + clk, 128);

    // update the status of the clk
    clkOn[clk] = 0;
}


/****************************************************************************
 * Set the power output for each output independently
 ***************************************************************************/
void Si5351mcu::setPower(uint8_t clk, uint8_t power) {
    // set the power to the correct var
    clkpower[clk] = power;

    // now enable the output to get it applied
    enable(clk);
}


/****************************************************************************
 * Private function to send the register data to the Si5351, arduino way.
 ***************************************************************************/
void Si5351mcu::i2cWrite(uint8_t regist, uint8_t value) {
    Wire.beginTransmission(SIADDR);
    Wire.write(regist);
    Wire.write(value);
    Wire.endTransmission();
}
